system/security/LdapSecurity/ldapconnection.cpp

```diff
@@ -1121,30 +1121,40 @@ class CLdapClient : public CInterface, implements ILdapClient
     virtual void init(IPermissionProcessor* pp)
     {
         m_pp = pp;
-        if(m_ldapconfig->getServerType() == OPEN_LDAP)
+        static CriticalSection  lcCrit;
+        static bool createdOU = false;
+        if (!createdOU)
         {
-            try
-            {
-                addDC(m_ldapconfig->getBasedn());
-            }
-            catch(...)
-            {
-            }
-            try
-            {
-                addGroup("Directory Administrators", m_ldapconfig->getBasedn());
-            }
-            catch(...)
+            CriticalBlock block(lcCrit);
+            if (!createdOU)
```

Comment:
This "double test" paradigm - with one test outside the critsec - is not technically safe on modern compilers/architectures (though we use it elsewhere and have not been bitten yet).

@ghalliday Do we have a preferred way to perform this type of operation safely for new code?